### ISSUE TITLE
Upcasting not working correctly with not exported derived classes

### ISSUE THREAD

**Issue by ChrLackner** [pybind/pybind11#1621]
## Issue description

Hi, the upcast of the return value of a function is not working as expected, if the return value is the holder type of a derived class, that is not exported, of an exported derived class (see code example below, I think then it is easier to understand). One would expect the Python type to always be `Derived`, but if there is a `shared_ptr<Base>` holding a `DerivedNotExported` instance, the python type is `Base` instead.
To reproduce by letting a function return a `shared_ptr<Base>` holding a `DerivedNotExported` object.

## Reproducible example code

C++ code:
``` cpp
#include <memory>
#include <pybind11/pybind11.h>

namespace py = pybind11;
using namespace std;

class Base
{
public:
  virtual ~Base() {}
};

class Derived : public Base
{
public:
  virtual ~Derived() {}
};

class DerivedNotExported : public Derived
{
public:
  virtual ~DerivedNotExported() {}
};

PYBIND11_MODULE(upcasting, m)
{
  py::class_<Base, shared_ptr<Base>>(m,"Base");
  py::class_<Derived, shared_ptr<Derived>, Base>(m, "Derived")
    .def(py::init([](bool exported) -> shared_ptr<Derived>
                  {
                    if(exported)
                      return make_shared<Derived>();
                    return make_shared<DerivedNotExported>();
                  }));

  m.def("GetDerived", [](bool exported) -> shared_ptr<Base>
        {
          if(exported)
            return make_shared<Derived>();
          return make_shared<DerivedNotExported>();
        });
}
```

Python code:
``` python
from upcasting import *

a = Derived(True)
b = Derived(False)

print(type(a))
print(type(b))

print(type(GetDerived(True)))
print(type(GetDerived(False)))
```
Gives
```
<class 'upcasting.Derived'>
<class 'upcasting.Derived'>
<class 'upcasting.Derived'>
<class 'upcasting.Base'>
```

Best
Christopher

**Comment by bstaletic** [comment]
This is working as expected.

In `Derived(True)`, you construct `Derived` and return it. Nothing surprising here.
In `Derived(False)`, you construct `DerivedNotExported` and return `Derived`. You get `Derived`
In `GetDerived(True)`, you construct `Derived`, return `Base` and get `Derived` because, if it is possible, pybind will return the most derived object it knows about.
In `GetDerived(False)`, you construct `DerivedNotExported`, and return `Base`. You get `Base`, because pybind doesn't know about `DerivedNotExported`.

**Comment by YannickJadoul** [comment]
I agree it would be nice to get `upcasting.Derived`in the last case, but that's technically expensive, quite hard, (and maybe impossible, I don't know). The issue is that you'd need to traverse the whole inheritance structure. Now pybind11 checks the most-derived type at runtime (like in Python), and if not known, it falls back to the static type (like C++).

**Comment by ChrLackner** [comment]
> In `GetDerived(True)`, you construct `Derived`, return `Base` and get `Derived` because, if it is possible, pybind will return the most derived object it knows about.
> In `GetDerived(False)`, you construct `DerivedNotExported`, and return `Base`. You get `Base`, because pybind doesn't know about `DerivedNotExported`.

As you said, the "perfect" thing would be if pybind returns the most derived object it knows about. In the last case this does not work, since it could return as `upcasting.Derived`. I see the point about being not easy - another obstacle would be if multiple inheritance or stuff is involved, then I guess it would get a full blown mess.. ;)

The problem why I was reporting this is the case if you have a Base as a member in another cpp class. If you put an instance of derivednotexported (cpp type) but derived (python type) in that class and get it out afterwards to python again it will just be an instance of base and for example stuff like pickling may call the wrong functions... This can lead to some very strange behavior...

So in short: if you implement pickling for another class that has base as its member and you pickle it by pickling base, the pickle implementation of base will be called and not the one from derived.

There are probably a lot of other cases that can lead to similar behavior.

I think to automatically traverse the inheritance structure all the time is not efficient, but maybe some classes can implement this, with an additional flag like `py::dynamic_attr` ?
Maybe even something like an explicit function (or return policy) like `py::cast_most_derived` would be enough that can be used to traverse threw registered types would be enough to help with that?

**Comment by bstaletic** [comment]
> As you said, the "perfect" thing would be if pybind returns the most derived object it knows about. In the last case this does not work, since it could return as upcasting.Derived.

Here's the problem with that. What if `DerivedNotExported` ingerited directly from `Base` and only `Base`? On top of that, what if `DerivedNotExported` was binary incompatible with `Derived`? Pybind can't know whether casting `Base*` to `Derived*` is safe, since it knows *nothing* about `DerivedNotExported`.

**Comment by ChrLackner** [comment]
Hm it could so by trying dynamic casting... But I guess this is too much of an overhead...
Hm would it be possible to export classes in a hidden way? So that `DerivedNotExported` is known in pybind and that it's base class is `Derived`, but the user only sees an instance of `Derived`? We have a range of classes in cpp (templated) that we do not want the user to see and confuse with... With the pybind constructors being able to create instances of derived classes this is super nicely possible, but if it is then upcasted to the base as soon as it is put to cpp and back again it is not really safe...

**Comment by YannickJadoul** [comment]
> I see the point about being not easy - another obstacle would be if multiple inheritance or stuff is involved, then I guess it would get a full blown mess.. ;)

Exactly; I don't immediately see how to do this.

> The problem why I was reporting this is the case if you have a Base as a member in another cpp class. If you put an instance of derivednotexported (cpp type) but derived (python type) in that class and get it out afterwards to python again it will just be an instance of base and for example stuff like pickling may call the wrong functions... This can lead to some very strange behavior... And I had this now in multiple cases and took some time to debug ;)

This is an interesting,cast, though. I guess the pickling should then be implemented as a virtual thing on the C++ object?

> Maybe even something like an explicit function (or return policy) like `py::cast_most_derived` would be enough that can be used to traverse threw registered types would be enough to help with that?

I wouldn't mind, but again, I don't know how to traverse a C++ hierarchy? The best I can currently think of is using a virtual member function on the type hierarchy itself `virtual py::object castMostDerived()` or so, that you would override yourself. But that's an intrusive approach and I don't think there's anything pybind11 can do about that?

**Comment by bstaletic** [comment]
> Hm it could so by trying dynamic casting... But I guess this is too much of an overhead...

I don't like the idea of `dynamic_cast`-ing "blindly", hoping something will work. Reminder that `dynamic_cast` also supports "side-casting", which could lead to even more surprising behaviour.

> Hm would it be possible to export classes in a hidden way?

Try making bindings with class names set to `""`. That would hide the class pretty well, though that too has some weird side-effects.

> The problem why I was reporting this is the case if you have a Base as a member in another cpp class. If you put an instance of derivednotexported (cpp type) but derived (python type) in that class and get it out afterwards to python again it will just be an instance of base and for example stuff like pickling may call the wrong functions... This can lead to some very strange behavior... And I had this now in multiple cases and took some time to debug ;)

Isn't this solved by virtual function calls? It doesn't matter what's the static type of an object, but what is its dynamic type.

**Comment by YannickJadoul** [comment]
> Hm it could so by trying dynamic casting... But I guess this is too much of an overhead...

Same here: the problem with dynamic casting is that you need to know the possible types to dynamic cast to. As far as I know, there's no way (other than virtual functions) to ask C++ "please traverse the hierarchy"?

That being said: if there is an approach, I'd be very curious to see it, and maybe it could be added as an optional thing pybind11 does?

**Comment by YannickJadoul** [comment]
> Isn't this solved by virtual function calls? It doesn't matter what's the static type of an object, but what is its dynamic type.

@bstaletic It doesn't work if you bind a lambda to Python that expects things to "just work".

**Comment by ChrLackner** [comment]
I agree that blindly dynamic casting threw the python class tree is not the solution. To explain the problem with pickling further:
1st: I do not want to inject python code into the cpp parts, as they should run standalone without python as well.
2nd: We have cpp serialization for efficient serialization of all the cpp stuff like arrays,... that we use when we pickle large objects. But we want to pickle large objects only once (first because of memory, but also so that they are the same object after unpickling again). So if we pickle an object that has other objects that may exist in python we basically give them to python again to be pickled and the python pickler decides if it is already pickled or not.
This is necessary because on a call to the pickle function we don't know if the member object is or is not pickled as the user may call something like
```
pickle.dumps([obj1, obj2])
```
with obj2 having obj1 as a member.

Because of this virtual functions are not enough and the correct python pickle function must be called.

**Comment by YannickJadoul** [comment]
>  1st: I do not want to inject python code into the cpp parts, as they should run standalone without python as well.

> Because of this virtual functions are not enough and the correct python pickle function must be called.

Agreed, it would be nice. But I don't know immediately how to do it in a general way (not even in a non-efficient, hierarchy traversing way). Do you have a suggestion?

The only reason casting to the actual dynamic type works is because of some dark `dynamic_cast<void*>` magic (well, not per se dark magic, but a reasonably unknown corner of the C++ standard, I'd say).

**Comment by YannickJadoul** [comment]
Oh, pybind11 does allow you intercept this and do something custom; didn't know that: https://pybind11.readthedocs.io/en/stable/advanced/classes.html#custom-automatic-downcasters

So that's probably the answer.

**Comment by ChrLackner** [comment]
This is new to me, looks exactly like what I need :) Thanks!

**Comment by YannickJadoul** [comment]
> This is new to me, looks exactly like what I need :) Thanks!

Same for me,actually!
There's even an optional template argument in this `polymorphic_type_hook` that allows for SFINAE, so you don't have to specialize on every single class of your hierarchy :-)
